### DISCUSSION TITLE
Android Native Library (.so) in MAUI Class Library is not embedded in .DLL anymore

### DISCUSSION THREAD

**Discussion by jurganson** [dotnet/maui#18459] [General] [answered]
I have a Xamarin.Android class library which I used to build to a .dll and import into my Xamarin project. However changing to MAUI I am now converting the class libraries to MAUI as well.

Inside these class libraries I have some different native android libraries (.so files) inside their respective ABI's. When building the .dll they used to get embedded inside the .dll when using the NativeAndroidLibrary build action. But now it seems like they are packaged in a .aar package instead and put beside the built .dll. Is there a way for me to embed the library files in the .dll, such that I don't have to supply the .aar as well?

If not, what are my options? I could the libraries in my MAUI app directly, or can I simply consume the AAR file in the MAUI app as well?

**Comment by nathenxbrewer** [comment]
Running into this issue as well. Did you ever figure this out?

**Reply by jurganson** [reply, marked as answer]
@nathenxbrewer Yeah kind of - I had to simply make sure to import the relevant .so files in my main solution as `<AndroidNativeLibrary>` following the ABI scheme, so:
```
- Android Solution
  - lib
    - arm64-v8a
      - mysofile.so
    - armeabi-v7a
      - mysofile.so
    - x86_64
      - mysofile.so
```

That did the trick for me. You might also be able to import the build AAR file as an `AndroidLibrary` instead, making sure to set `Bind="False"`, with the prerequisite that you did the binding in the other project already and just want the library. But I have not tested that.

**Comment by jurganson** [comment]
I am closing this discussion, please see the answer in @nathenxbrewer comment reply.

**Comment by bbhxwl** [comment]
I don't know how to bind an aar, can you help me?

**Reply by jurganson** [reply]
Binding a Java library can potentially be quite a process and each library requires different tinkering and modifying, I would suggest hitting up Google and searching for resources, there are quite a few. If you can't find any good ones for MAUI, I would go ahead and search for Xamarin-related resources, as the process is largely still the same. Then you can simply follow the migration process provided by [Microsoft](https://learn.microsoft.com/en-us/dotnet/maui/migration/android-binding-projects?view=net-maui-8.0)

A really good Xamarin resource is [this article by John Thiriet](https://johnthiriet.com/xamarin-android-binding-library-preparation/)

Good luck on your binding project!
